File: Seccion 16 - Arboles en C++/Seccion_16_Python/01_insertar_nodos.py

class Nodo:
    
    def __init__(self, dato=None, izq=None, der=None):
        self.dato = dato
        self.izq = izq
        self.der = der
        self.padre = None
    
    def __str__(self):
        return "{} -> ".format(self.dato)

class Arbol:
    
    def __init__(self):
        self.raiz = None
    
    def agregar_nodo(self, nodo: Nodo):
        if self.raiz == None:
            self.raiz = nodo
        else:
            aux = self.raiz
            padre = None
            
            while aux != None:
                padre = aux
                if nodo.dato >= aux.dato:
                    aux = aux.der
                else:
                    aux = aux.izq
            
            if nodo.dato >= padre.dato:
                padre.der = nodo
            else:
                padre.izq = nodo

def menu():
    n: int
    opcion: int = 0
    
    arbol = Arbol()
    
    while opcion != 2:
        print("\n.:MENU:.")
        print("1. Insertar un nuevo nodo.")
        print("2. Salir.")
        opcion = int(input("Opción: "))
        
        if opcion == 1:
            n = int(input("\nDigite un número entero: "))
            nodo = Nodo(dato=n)
            arbol.agregar_nodo(nodo)

def main():
    menu()

if __name__ == "__main__":
    main()

File: Seccion 16 - Arboles en C++/Seccion_16_Dart/01_insertar_nodos.dart

import 'dart:io';

class Nodo {

  int dato;
  Nodo izq;
  Nodo der;
  Nodo padre;

  Nodo(this.dato);

  String str() {
    return "$dato -> ";
  }
}

class Arbol {

  Nodo raiz;

  Arbol();

  void agregarNodo(Nodo nodo) {
    if (raiz == null) {
      raiz = nodo;
    } else {
      Nodo aux = raiz;
      Nodo padre;

      while (aux != null) {
        padre = aux;
        if (nodo.dato >= aux.dato) {
          aux = aux.der;
        } else {
          aux = aux.izq;
        }

        nodo.padre = padre;
        if (nodo.dato >= padre.dato) {
          padre.der = nodo;
        } else {
          padre.izq = nodo;
        }
      }
    }
  }
}

void menu() {
  int n;
  int opcion = 0;
  Nodo nodo;

  Arbol arbol = new Arbol();

  do {
    stdout.writeln("\t.:MENU:.");
    stdout.writeln("1. Insertar un nuevo nodo.");
    stdout.writeln("2. Salir.");
    stdout.write("Opción: ");
    opcion = int.parse(stdin.readLineSync());

    switch (opcion) {
      case 1:
        stdout.write("\nDigite un número entero: ");
        n = int.parse(stdin.readLineSync());
        nodo = new Nodo(n);
        arbol.agregarNodo(nodo);
        break;
      
      default:
        break;
    }
  } while (opcion != 2);
}

void main() {

  menu();

}